File: crates/quilltap-core/src/model_classes.rs
//! Port of v4's lib/llm/model-classes.ts — the built-in LLM capability tiers
//! and the two lookups over them. Pure constant data + table search.

/// A model class: a named capability tier a connection profile can reference.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ModelClass {
    /// Unique identifier / display name.
    pub name: &'static str,
    /// Single-letter tier designation (A = smallest, D = largest).
    pub tier: &'static str,
    /// Maximum context window size in tokens.
    pub max_context: i64,
    /// Maximum output/completion size in tokens.
    pub max_output: i64,
    /// Capability tags for categorization and filtering.
    pub tags: &'static [&'static str],
    /// Quality ranking (0 = lowest, higher = better).
    pub quality: i64,
}

/// Built-in model classes defining standard LLM capability tiers. Order and
/// values mirror v4's `MODEL_CLASSES` exactly.
pub const MODEL_CLASSES: &[ModelClass] = &[
    ModelClass {
        name: "Compact",
        tier: "A",
        max_context: 32000,
        max_output: 4000,
        tags: &["SMALL", "CHEAP", "LOCAL"],
        quality: 0,
    },
    ModelClass {
        name: "Standard",
        tier: "B",
        max_context: 128000,
        max_output: 16000,
        tags: &["BUDGET"],
        quality: 1,
    },
    ModelClass {
        name: "Extended",
        tier: "C",
        max_context: 200000,
        max_output: 128000,
        tags: &["CREATIVE", "THINKING"],
        quality: 2,
    },
    ModelClass {
        name: "Deep",
        tier: "D",
        max_context: 1000000,
        max_output: 128000,
        tags: &["MAX"],
        quality: 3,
    },
];

/// Look up a model class by exact name, or `None` if unknown.
pub fn get_model_class(name: &str) -> Option<&'static ModelClass> {
    MODEL_CLASSES.iter().find(|mc| mc.name == name)
}

/// Whether `name` matches a known model class (exact, case-sensitive).
pub fn is_valid_model_class_name(name: &str) -> bool {
    MODEL_CLASSES.iter().any(|mc| mc.name == name)
}

File: apps/web/src/app/screens/settings/providers/model-classes.ts
/**
 * Model classes — pure constant capability tiers (v4 `lib/llm/model-classes.ts`).
 * A connection profile may reference one by name.
 */
export interface ModelClass {
  name: string;
  tier: string;
  maxContext: number;
  maxOutput: number;
  tags: readonly string[];
  quality: number;
}

export const MODEL_CLASSES: readonly ModelClass[] = [
  {
    name: 'Compact',
    tier: 'A',
    maxContext: 32000,
    maxOutput: 4000,
    tags: ['SMALL', 'CHEAP', 'LOCAL'],
    quality: 0,
  },
  {
    name: 'Standard',
    tier: 'B',
    maxContext: 128000,
    maxOutput: 16000,
    tags: ['BUDGET'],
    quality: 1,
  },
  {
    name: 'Extended',
    tier: 'C',
    maxContext: 200000,
    maxOutput: 128000,
    tags: ['CREATIVE', 'THINKING'],
    quality: 2,
  },
  { name: 'Deep', tier: 'D', maxContext: 1000000, maxOutput: 128000, tags: ['MAX'], quality: 3 },
];

export function getModelClass(name: string): ModelClass | undefined {
  return MODEL_CLASSES.find((mc) => mc.name === name);
}

/** Normalize a profile name for uniqueness (v4 `normalizeProfileName`): trim + lower. */
export function normalizeProfileName(name: string): string {
  return name.trim().toLowerCase();
}

/** `desired`, else `desired (2)`, … until unique vs `taken` (v4 `makeUniqueProfileName`). */
export function makeUniqueProfileName(desired: string, taken: Set<string>): string {
  const base = desired.trim();
  if (!taken.has(normalizeProfileName(base))) {
    return base;
  }
  for (let n = 2; ; n++) {
    const candidate = `${base} (${n})`;
    if (!taken.has(normalizeProfileName(candidate))) {
      return candidate;
    }
  }
}
